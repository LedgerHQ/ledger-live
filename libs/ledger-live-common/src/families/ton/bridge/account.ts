import { defaultUpdateTransaction } from "@ledgerhq/coin-framework/bridge/jsHelpers";
import { patchOperationWithHash } from "@ledgerhq/coin-framework/operation";
import {
  AmountRequired,
  InvalidAddress,
  InvalidAddressBecauseDestinationIsAlsoSource,
  NotEnoughBalance,
  RecipientRequired,
} from "@ledgerhq/errors";
import {
  Account,
  AccountBridge,
  AccountLike,
  BroadcastFnSignature,
  SignOperationEvent,
  SignOperationFnSignature,
} from "@ledgerhq/types-live";
import { TonTransport } from "@ton-community/ton-ledger";
import BigNumber from "bignumber.js";
import { Observable } from "rxjs";
import { getMainAccount } from "../../../account/helpers";
import { makeAccountBridgeReceive, makeSync } from "../../../bridge/jsHelpers";
import { withDevice } from "../../../hw/deviceAccess";
import { TonCommentInvalid } from "../errors";
import { TonOperation, Transaction, TransactionStatus } from "../types";
import {
  addressesAreEqual,
  commentIsValid,
  getAddress,
  getLedgerTonPath,
  getTonEstimatedFees,
  isAddressValid,
  packTransaction,
  transactionToHwParams,
} from "../utils";
import { getAccountShape } from "./bridgeHelpers/accountShape";
import { broadcastTx, fetchAccountInfo } from "./bridgeHelpers/api";

const estimateMaxSpendable = async ({
  account,
  parentAccount,
  transaction,
}: {
  account: AccountLike;
  parentAccount?: Account | null | undefined;
  transaction?: Transaction | null | undefined;
}): Promise<BigNumber> => {
  const a = getMainAccount(account, parentAccount);
  let balance = a.spendableBalance;

  if (balance.eq(0)) return balance;

  const accountInfo = await fetchAccountInfo(getAddress(a).address);
  const estimatedFees = transaction
    ? transaction.fees ??
      (await getTonEstimatedFees(
        a,
        accountInfo.status === "uninit",
        transactionToHwParams(transaction, accountInfo.seqno),
      ))
    : BigNumber(0);

  if (balance.lte(estimatedFees)) return new BigNumber(0);

  balance = balance.minus(estimatedFees);

  return balance;
};

const createTransaction = (): Transaction => {
  return {
    family: "ton",
    amount: new BigNumber(0),
    fees: new BigNumber(0),
    recipient: "",
    useAllAmount: false,
    comment: {
      isEncrypted: false,
      text: "",
    },
  };
};

const getTransactionStatus = async (a: Account, t: Transaction): Promise<TransactionStatus> => {
  const errors: TransactionStatus["errors"] = {};
  const warnings: TransactionStatus["warnings"] = {};

  const { balance, spendableBalance } = a;
  const { address } = getAddress(a);
  const { recipient, useAllAmount } = t;
  let { amount } = t;

  if (!recipient) {
    errors.recipient = new RecipientRequired();
  } else if (!isAddressValid(recipient)) {
    errors.recipient = new InvalidAddress("", {
      currencyName: a.currency.name,
    });
  } else if (addressesAreEqual(address, recipient)) {
    errors.recipient = new InvalidAddressBecauseDestinationIsAlsoSource();
  }

  if (!isAddressValid(address)) {
    errors.sender = new InvalidAddress("", {
      currencyName: a.currency.name,
    });
  }

  const estimatedFees = t.fees;

  let totalSpent = BigNumber(0);

  if (useAllAmount) {
    totalSpent = spendableBalance;
    amount = totalSpent.minus(estimatedFees);
    if (amount.lte(0) || totalSpent.gt(balance)) {
      errors.amount = new NotEnoughBalance();
    }
  } else {
    totalSpent = amount.plus(estimatedFees);
    if (totalSpent.gt(spendableBalance)) {
      errors.amount = new NotEnoughBalance();
    }
    if (amount.eq(0)) {
      errors.amount = new AmountRequired();
    }
  }

  if (t.comment.isEncrypted || !commentIsValid(t.comment)) {
    errors.comment = new TonCommentInvalid();
  }

  return {
    errors,
    warnings,
    estimatedFees,
    amount,
    totalSpent,
  };
};

const prepareTransaction = async (a: Account, t: Transaction): Promise<Transaction> => {
  const accountInfo = await fetchAccountInfo(getAddress(a).address);
  const fees = await getTonEstimatedFees(
    a,
    accountInfo.status === "uninit",
    transactionToHwParams(t, accountInfo.seqno),
  );

  const amount = t.useAllAmount ? a.spendableBalance.minus(t.fees) : t.amount;

  return defaultUpdateTransaction(t, { fees, amount });
};

const sync = makeSync({
  getAccountShape,
  postSync: (_, a) => {
    const operations = a.operations || [];
    const initialPendingOps = a.pendingOperations || [];
    const pendingOperations = initialPendingOps.filter(
      pOp => !operations.some(o => o.id === pOp.id),
    );
    return { ...a, pendingOperations };
  },
});

const receive = makeAccountBridgeReceive();

const signOperation: SignOperationFnSignature<Transaction> = ({
  account,
  deviceId,
  transaction,
}): Observable<SignOperationEvent> =>
  withDevice(deviceId)(
    transport =>
      new Observable(o => {
        async function main() {
          // log("debug", "[signOperation] start fn");

          const { recipient, amount, fees, comment } = transaction;
          const { address, derivationPath } = getAddress(account);
          const accountInfo = await fetchAccountInfo(address);

          const app = new TonTransport(transport);

          o.next({
            type: "device-signature-requested",
          });

          // Sign by device
          // it already verifies the signature inside
          const sig = await app.signTransaction(
            getLedgerTonPath(derivationPath),
            transactionToHwParams(transaction, accountInfo.seqno),
          );

          o.next({
            type: "device-signature-granted",
          });

          const signature = packTransaction(account, accountInfo.status === "uninit", sig);
          const hash = sig.hash().toString("hex");

          const operation: TonOperation = {
            // we'll patch operation when broadcasting
            id: hash,
            hash,
            type: "OUT",
            senders: [address],
            recipients: [recipient],
            accountId: account.id,
            value: amount.plus(fees),
            fee: fees,
            blockHash: null,
            blockHeight: null,
            date: new Date(),
            extra: {
              // we don't know yet, will be patched in final operation
              lt: "",
              explorerHash: "",
              comment: comment,
            },
          };

          o.next({
            type: "signed",
            signedOperation: {
              operation,
              signature,
            },
          });
        }

        main().then(
          () => o.complete(),
          e => o.error(e),
        );
      }),
  );

const broadcast: BroadcastFnSignature = async ({ signedOperation: { signature, operation } }) => {
  const hash = await broadcastTx(signature);
  return patchOperationWithHash(operation, hash);
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction: defaultUpdateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export { accountBridge };
