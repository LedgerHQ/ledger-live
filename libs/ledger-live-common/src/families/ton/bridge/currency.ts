import { CurrencyBridge } from "@ledgerhq/types-live";
import { makeScanAccounts } from "../../../bridge/jsHelpers";
import { getAccountShape } from "./bridgeHelpers/accountShape";

const scanAccounts = makeScanAccounts({ getAccountShape });

export const currencyBridge: CurrencyBridge = {
  preload: () => Promise.resolve({}),
  hydrate: () => {},
  scanAccounts,
};
