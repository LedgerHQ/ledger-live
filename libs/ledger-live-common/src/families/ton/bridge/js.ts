import { accountBridge } from "./account";
import { currencyBridge } from "./currency";

export default {
  currencyBridge,
  accountBridge,
};
