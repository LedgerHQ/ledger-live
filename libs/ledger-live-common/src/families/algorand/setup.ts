// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-algorand/bridge/js";
import makeCliTools from "@ledgerhq/coin-algorand/cli-transaction";
import algorandResolver from "@ledgerhq/coin-algorand/hw-getAddress";
import { Transaction } from "@ledgerhq/coin-algorand/types";
import Algorand from "@ledgerhq/hw-app-algorand";
import Transport from "@ledgerhq/hw-transport";
import network from "@ledgerhq/live-network/network";
import type { Bridge } from "@ledgerhq/types-live";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Algorand> = (transport: Transport) => {
  return new Algorand(transport);
};

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), network);

const resolver: Resolver = createResolver(createSigner, algorandResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
