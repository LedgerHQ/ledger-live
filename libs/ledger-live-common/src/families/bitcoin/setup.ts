// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { firstValueFrom, from } from "rxjs";
import Transport from "@ledgerhq/hw-transport";
import { Bridge } from "@ledgerhq/types-live";
import { CryptoCurrency } from "@ledgerhq/types-cryptoassets";
import Btc from "@ledgerhq/hw-app-btc";
import { createBridges } from "@ledgerhq/coin-bitcoin/bridge/js";
import type { SignerContext } from "@ledgerhq/coin-bitcoin/signer";
import makeCliTools from "@ledgerhq/coin-bitcoin/cli-transaction";
import bitcoinResolver from "@ledgerhq/coin-bitcoin/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-bitcoin/hw-signMessage";
import { Transaction } from "@ledgerhq/coin-bitcoin/types";
import { GetAddressOptions, Resolver } from "../../hw/getAddress/types";
import { withDevice } from "../../hw/deviceAccess";
import { startSpan } from "../../performance";
import { GetAddressFn } from "@ledgerhq/coin-framework/bridge/getAddressWrapper";
import { getCurrencyConfiguration } from "../../config";
import { BitcoinConfigInfo } from "@ledgerhq/coin-bitcoin/lib/config";
import { SignMessage } from "../../hw/signMessage/types";

const createSigner = (transport: Transport, currency: CryptoCurrency) => {
  return new Btc({ transport, currency: currency.id });
};

const signerContext: SignerContext = <T>(
  deviceId: string,
  crypto: CryptoCurrency,
  fn: (signer: Btc) => Promise<T>,
): Promise<T> =>
  firstValueFrom(
    withDevice(deviceId)((transport: Transport) => from(fn(createSigner(transport, crypto)))),
  );

const getCurrencyConfig = (currency: CryptoCurrency) => {
  return { info: getCurrencyConfiguration<BitcoinConfigInfo>(currency) };
};

const perfLogger = {
  startSpan,
};

const bridge: Bridge<Transaction> = createBridges(signerContext, perfLogger, getCurrencyConfig);

export function createMessageSigner(): SignMessage {
  return (transport, account, messageData) => {
    const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
    return signMessage(signerContext)("", account, messageData);
  };
}

const messageSigner = {
  signMessage: createMessageSigner(),
};

const resolver: Resolver = (
  transport: Transport,
  addressOpt: GetAddressOptions,
): ReturnType<GetAddressFn> => {
  const signerContext: SignerContext = (_, crypto, fn) => fn(createSigner(transport, crypto));
  return bitcoinResolver(signerContext)("", addressOpt);
};

const cliTools = makeCliTools();

export { bridge, cliTools, resolver, messageSigner, signerContext };
