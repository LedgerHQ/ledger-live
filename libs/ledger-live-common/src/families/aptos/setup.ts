// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import {
  AptosAccount,
  TransactionStatus,
  createBridges,
  type Transaction,
} from "@ledgerhq/coin-aptos";
import Transport from "@ledgerhq/hw-transport";
import Aptos from "@ledgerhq/hw-app-aptos";
import type { Bridge } from "@ledgerhq/types-live";
import aptosResolver from "@ledgerhq/coin-aptos/signer/index";
import makeCliTools from "@ledgerhq/coin-aptos/test/cli";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<Aptos> = (transport: Transport) => {
  return new Aptos(transport);
};

const bridge: Bridge<Transaction, AptosAccount, TransactionStatus> = createBridges(
  executeWithSigner(createSigner),
);

const resolver: Resolver = createResolver(createSigner, aptosResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
