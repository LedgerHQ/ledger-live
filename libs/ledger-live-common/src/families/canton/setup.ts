// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-canton/bridge/index";
import { getEnv } from "@ledgerhq/live-env";
import Transport from "@ledgerhq/hw-transport";
import { CantonSigner } from "@ledgerhq/coin-canton";
import cantonResolver from "@ledgerhq/coin-canton/signer";
import type { Account, Bridge } from "@ledgerhq/types-live";
import makeCliTools from "@ledgerhq/coin-canton/test/cli";
import { CantonCoinConfig } from "@ledgerhq/coin-canton/config";
import { TransactionStatus, Transaction } from "@ledgerhq/coin-canton/types";
import { LegacySignerCanton } from "@ledgerhq/live-signer-canton";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import cantonBridgeMock from "./bridge/mock";
import { Resolver } from "../../hw/getAddress/types";
import { getCurrencyConfiguration } from "../../config";
import { getCryptoCurrencyById } from "../../currencies";

const createSigner: CreateSigner<CantonSigner> = (transport: Transport) => {
  return new LegacySignerCanton(transport);
};

const getCurrencyConfig = () => {
  // Use devnet for development/testing
  const currencyId = getEnv("MOCK") ? "canton_network_devnet" : "canton_network_devnet";
  return getCurrencyConfiguration<CantonCoinConfig>(getCryptoCurrencyById(currencyId));
};

const bridge: Bridge<Transaction, Account, TransactionStatus> = getEnv("MOCK")
  ? cantonBridgeMock
  : createBridges(executeWithSigner(createSigner), getCurrencyConfig);

const resolver: Resolver = createResolver(createSigner, cantonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, resolver };
