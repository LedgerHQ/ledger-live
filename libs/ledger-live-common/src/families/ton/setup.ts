// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-ton/bridge/js";
import makeCliTools from "@ledgerhq/coin-ton/cli-transaction";
import tonResolver from "@ledgerhq/coin-ton/hw-getAddress";
import { signMessage } from "@ledgerhq/coin-ton/hw-signMessage";
import { TonCoinConfig } from "@ledgerhq/coin-ton/lib/config";
import { TonSigner } from "@ledgerhq/coin-ton/lib/signer";
import { Transaction } from "@ledgerhq/coin-ton/types";
import { getCryptoCurrencyById } from "@ledgerhq/cryptoassets/currencies";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import { TonTransport as Ton } from "@ton-community/ton-ledger";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import { getCurrencyConfiguration } from "../../config";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<TonSigner> = (transport: Transport) => new Ton(transport);

const getCoinConfig: TonCoinConfig = () =>
  getCurrencyConfiguration<ReturnType<TonCoinConfig>>(getCryptoCurrencyById("ton"));

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner), getCoinConfig);

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, tonResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
