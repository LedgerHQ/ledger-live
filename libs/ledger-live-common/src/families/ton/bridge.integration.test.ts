import { dataset } from "@ledgerhq/coin-ton/bridge.integration.test";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import "../../__tests__/test-helpers/setup";

testBridge(dataset);
