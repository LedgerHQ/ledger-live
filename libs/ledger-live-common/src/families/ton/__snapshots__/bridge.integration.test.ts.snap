// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ton currency bridge scanAccounts ton seed 1 1`] = `
[
  {
    "balance": "1000000000",
    "currencyId": "ton",
    "derivationMode": "ton",
    "freshAddress": "UQCOvQLYvTcbi5tL9MaDNzuVl3-J3vATimNm9yO5XPafLfV4",
    "freshAddressPath": "44'/607'/0'/0'/0'/0'",
    "freshAddresses": [
      {
        "address": "UQCOvQLYvTcbi5tL9MaDNzuVl3-J3vATimNm9yO5XPafLfV4",
        "derivationPath": "44'/607'/0'/0'/0'/0'",
      },
    ],
    "id": "js:2:ton:86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060:ton",
    "index": 0,
    "name": "TON 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": [],
    "seedIdentifier": "86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060",
    "spendableBalance": "1000000000",
    "starred": false,
    "swapHistory": [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": true,
    "xpub": "86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060",
  },
  {
    "balance": "0",
    "currencyId": "ton",
    "derivationMode": "ton",
    "freshAddress": "UQAui6M4jOYOezUGfmeONA22Ars9yjd34YIGdAR1Pcpp4sgR",
    "freshAddressPath": "44'/607'/0'/0'/1'/0'",
    "freshAddresses": [
      {
        "address": "UQAui6M4jOYOezUGfmeONA22Ars9yjd34YIGdAR1Pcpp4sgR",
        "derivationPath": "44'/607'/0'/0'/1'/0'",
      },
    ],
    "id": "js:2:ton:b5177c2b32f9d72fa8c673cc3d61acec6a9f68eb5e4945445fdbb48a45eb4887:ton",
    "index": 1,
    "name": "TON 2",
    "nfts": undefined,
    "operationsCount": 0,
    "pendingOperations": [],
    "seedIdentifier": "86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060",
    "spendableBalance": "0",
    "starred": false,
    "swapHistory": [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": false,
    "xpub": "b5177c2b32f9d72fa8c673cc3d61acec6a9f68eb5e4945445fdbb48a45eb4887",
  },
]
`;

exports[`ton currency bridge scanAccounts ton seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:ton:86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060:ton",
      "blockHash": null,
      "blockHeight": 35650189,
      "contract": undefined,
      "extra": {
        "comment": {
          "isEncrypted": false,
          "text": "",
        },
        "explorerHash": "mzCqNBjsEDmpgq5VMfNfNqnslW64f/B1seatUCej9TY=",
        "lt": "44161792000003",
      },
      "fee": "0",
      "hasFailed": false,
      "hash": "V0y/lJDCYzAgmzNVA+dapRQemNP+AJpVhCxBhs7eNfQ=",
      "id": "js:2:ton:86196cb40cd25e9e696bc808e3f2c074ce0b39f2a2a9d482a68eafef86e4a060:ton-V0y/lJDCYzAgmzNVA+dapRQemNP+AJpVhCxBhs7eNfQ=-IN",
      "operator": undefined,
      "recipients": [
        "UQCOvQLYvTcbi5tL9MaDNzuVl3-J3vATimNm9yO5XPafLfV4",
      ],
      "senders": [
        "UQDzd8aeBOU-jqYw_ZSuZjceI5p-F4b7HMprAsUJAtRPbMol",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "1000000000",
    },
  ],
  [],
]
`;
